### PR TITLE
🐛 amp-story-shopping Fixes remote shopping data loading on refresh with page attachment open

### LINKED ISSUE #38040: amp-story-shopping page attachment wth remote URL page attachment open before shopping data is loaded
### Description

Expected: shopping data will load properly regardless if the page attachment opens first or not.

### Reproduction Steps

Intermittent bug: Open page attachment with a remote URL, then refresh the page while the page attachment is still open

### PR DESCRIPTION
**Pull request by jshamble**
Closes #38040

Moved to https://github.com/ampproject/amphtml/pull/38035

### CODE REVIEW

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js [hunk 1]
```diff
@@ -157,15 +157,20 @@ export class AmpStoryShoppingAttachment extends AMP.BaseElement {
...
-      // This happens when a product card is clicked in the PLP template.
+      // This happens when a product card is clicked in the PLP template
+      // or when the shopping attachment is opened when product data is loaded.
       this.updateTemplate_(shoppingData);
-    } else {
+    }
+
+    if (!shoppingData.activeProductData || !this.isOnActivePage_()) {
+      return;
+    }
```
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:171]: Without checking if we're on the active page first, code will run on all pages with this component.

File: extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js [hunk 2]
```diff
@@ -157,15 +157,20 @@ export class AmpStoryShoppingAttachment extends AMP.BaseElement {
...
-    if (!shoppingData.activeProductData || !this.isOnActivePage_()) {
-      return;
-    }
     const isOpen = this.storeService_.get(StateProperty.PAGE_ATTACHMENT_STATE);
+
     if (isOpen) {
       // If open, update the template.
-      // This happens when a product card is clicked in the PLP template.
+      // This happens when a product card is clicked in the PLP template
+      // or when the shopping attachment is opened when product data is loaded.
```
Comment (processprocess) [inline extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js:165]:
```suggestion
      // or when page attachment state is open on refresh.
```

**Review by processprocess — CHANGES_REQUESTED** [review]
The conditionals are getting hard to follow and are spanning across 2 functions.
This makes the code difficult to read and will make maintenance more difficult.

To follow the call stack, it looks like `onAttachmentStateUpdate_` is  called on refresh before data is loaded, causing [Object.values here](https://github.com/ampproject/amphtml/blob/main/extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js#L198) to throw the error.

Is there a different approach that reduces conditionals?

**Comment by jshamble** [comment]
@processprocess couldn't think of a different approach that reduces the conditionals after trying a few combinations.

**Comment by processprocess** [comment]
Check out the approach in #38090 and how it uses `Promise.reject` to prevent `layoutCallback` from running.
This way can remove conditionals and log why the component isn't rendering.
The approach could also fix #38035

**Comment by gmajoulet** [comment]
Could you link me to the code that reopens the shopping attachment on page reload?
I'm curious as to why you need to move these checks, the `activeProductData` should be restored from the history entry, and everything should work properly

**Comment by jshamble** [comment]
@gmajoulet 
Shopping attachment updates happen in the following two functions here via the store service
```
onAttachmentStateUpdate_
onShoppingDataUpdate_
```

Specifically when `updateTemplate_` is called.

```
   this.storeService_.subscribe(
      StateProperty.PAGE_ATTACHMENT_STATE,
      (isOpen) => this.onAttachmentStateUpdate_(isOpen),
      true /** callToInitialize */
    );
    this.storeService_.subscribe(
      StateProperty.SHOPPING_DATA,
      (shoppingData) => this.onShoppingDataUpdate_(shoppingData),
      true /** callToInitialize */
    );
```

The issue is happening because things are loaded out of order - i.e. the page attachment is updated before all of the shopping data is loaded (specifically the remote data). 
I'll set some things up with promises so that everything is synced up properly.

**Comment by processprocess** [comment]
```
page attachment is updated before all of the shopping data is loaded (specifically the remote data).
```

I Think @gmajoulet is suggesting that the store service state should be restored and already have the shopping data. That's how other state properties are restored (like page and attachment state). It would be good to look into why it's not restoring and returning an empty object. Maybe it is restoring but since the example has an empty items data it's throwing the error?

**Comment by jshamble** [comment]
@processprocess 
I did file a bug for that a few days ago actually (the store service not restoring the data as intended), was going to look into it after this issue was resolved, but I can close that issue and just take care of it as part of this PR.
https://github.com/ampproject/amphtml/issues/38074

**Comment by jshamble** [comment]
Moved to https://github.com/ampproject/amphtml/pull/38035
